examples/cpp/topic_instances/SubscriberApp.cpp

```diff
@@ -0,0 +1,260 @@
+    if (subscriber_ == nullptr)
+    {
+        throw std::runtime_error("Subscriber initialization failed");
+    }
+
+    // Create the topic
+    TopicQos topic_qos = TOPIC_QOS_DEFAULT;
+    participant_->get_default_topic_qos(topic_qos);
+    topic_ = participant_->create_topic(config.topic_name, type_.get_type_name(), topic_qos);
+    if (topic_ == nullptr)
+    {
+        throw std::runtime_error("Topic initialization failed");
+    }
+
+    // Create the data reader
+    DataReaderQos reader_qos = DATAREADER_QOS_DEFAULT;
+    subscriber_->get_default_datareader_qos(reader_qos);
+    reader_qos.reliability().kind = ReliabilityQosPolicyKind::RELIABLE_RELIABILITY_QOS;
+    reader_qos.durability().kind = DurabilityQosPolicyKind::TRANSIENT_LOCAL_DURABILITY_QOS;
+    reader_qos.history().kind = HistoryQosPolicyKind::KEEP_LAST_HISTORY_QOS;
+    uint32_t sample_limit = samples_;
+    if (sample_limit == 0)
+    {
+        sample_limit = DATAREADER_QOS_DEFAULT.resource_limits().max_samples_per_instance;
+    }
+    reader_qos.resource_limits().max_instances = config.instances;
+    reader_qos.history().depth = sample_limit;
+    reader_qos.resource_limits().max_samples_per_instance = sample_limit;
+    reader_qos.resource_limits().max_samples = sample_limit * config.instances;
+    reader_ = subscriber_->create_datareader(topic_, reader_qos, this, StatusMask::all());
+    if (reader_ == nullptr)
+    {
+        throw std::runtime_error("DataReader initialization failed");
+    }
+}
+
+SubscriberApp::~SubscriberApp()
+{
+    if (nullptr != participant_)
+    {
+        // Delete DDS entities contained within the DomainParticipant
+        participant_->delete_contained_entities();
+
+        // Delete DomainParticipant
+        DomainParticipantFactory::get_instance()->delete_participant(participant_);
+    }
+}
+
+void SubscriberApp::on_subscription_matched(
+        DataReader* /*reader*/,
+        const SubscriptionMatchedStatus& info)
+{
+    if (info.current_count_change == 1)
+    {
+        std::cout << "Subscriber matched." << std::endl;
+    }
+    else if (info.current_count_change == -1)
+    {
+        std::cout << "Subscriber unmatched." << std::endl;
+    }
+    else
+    {
+        std::cout << info.current_count_change
+                  << " is not a valid value for SubscriptionMatchedStatus current count change" << std::endl;
+    }
+}
+
+void SubscriberApp::on_data_available(
+        DataReader* reader)
+{
+    SampleInfo info;
+    while ((!is_stopped()) && (RETCODE_OK == reader->take_next_sample(&shape_, &info)))
+    {
+        if ((info.instance_state == ALIVE_INSTANCE_STATE) && info.valid_data)
+        {
+            // Add the received sample to the corresponding instance handle counter
+            if (samples_per_instance_.count(info.instance_handle) > 0)
+            {
+                // Instance handle exists in the map, increase the counter
+                samples_per_instance_[info.instance_handle]++;
```

Comment:
I have tried to run the Subscriber with --samples and it doesn't stop after 10 samples. The problem can be in the way this is increased

Reply:
It works on my side, lets reproduce it on monday

Reply:
I have tried `subscriber --samples 5` and the subscriber doesn't stop after it received 5 samples each

Reply:
Fixed in [04292b7](https://github.com/eProsima/Fast-DDS/pull/5033/commits/04292b738bd260f102123627cbb47d273aa7209b)